### DISCUSSION TITLE
[BUG] Using list[bytes] with Annotated and choosing "Send empty value" in swagger UI, causes internal FastApi AttributeError

### DISCUSSION THREAD

**Discussion by plsholdmybeer** [fastapi/fastapi#9364] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated
from fastapi import FastAPI, File, Form, UploadFile


app = FastAPI()


@app.post("/files/")
async def create_file(
    files1_crash_if_empty: Annotated[list[bytes] | None, File(description="A file read as bytes")] = None,
    files1_works_no_empty: list[bytes] | None = File(description="A file read as bytes"),
    files2_works_ok_empty: Annotated[list[UploadFile] | None, File(description="A file read as UploadFile")] = None,
):
    a = {"file_sizes": [len(file) for file in files1_crash_if_empty]} if files1_crash_if_empty else None
    a = {"file_sizes": [len(file) for file in files1_works_no_empty]} if files1_works_no_empty else None
    b = {"filename": [file.filename for file in files2_works_ok_empty]} if files2_works_ok_empty else None
    return (a, b)
```

### Description

files1_works_no_empty works fine because there is no field "send empty value"
files2_works_ok_empty works fine when selecting empty

It appears that when using Annotated, field "send empty value" is being shown even though it should not. Leaving this here for review, I don't have enough knowledge on fastapi internals to create a pull request.

`files1_crash_if_empty` if submitted via swagger UI with selected "Send empty value", causes an internal exception in fastapi, traceback below

```
Traceback (most recent call last):
  File "...\.venv\lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 429, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "...\.venv\lib\site-packages\uvicorn\middleware\proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "...\.venv\lib\site-packages\fastapi\applications.py", line 276, in __call__
    await super().__call__(scope, receive, send)
  File "...\.venv\lib\site-packages\starlette\applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "...\.venv\lib\site-packages\starlette\middleware\errors.py", line 184, in __call__
    raise exc
  File "...\.venv\lib\site-packages\starlette\middleware\errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "...\.venv\lib\site-packages\starlette\middleware\exceptions.py", line 79, in __call__
    raise exc
  File "...\.venv\lib\site-packages\starlette\middleware\exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "...\.venv\lib\site-packages\fastapi\middleware\asyncexitstack.py", line 21, in __call__
    raise e
  File "...\.venv\lib\site-packages\fastapi\middleware\asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "...\.venv\lib\site-packages\starlette\routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "...\.venv\lib\site-packages\starlette\routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "...\.venv\lib\site-packages\starlette\routing.py", line 66, in app
    response = await func(request)
  File "...\.venv\lib\site-packages\fastapi\routing.py", line 219, in app
    solved_result = await solve_dependencies(
  File "...\.venv\lib\site-packages\fastapi\dependencies\utils.py", line 649, in solve_dependencies
    ) = await request_body_to_args(  # body_params checked above
  File "...\.venv\lib\site-packages\fastapi\dependencies\utils.py", line 781, in request_body_to_args       
    tg.start_soon(process_fn, sub_value.read)
AttributeError: 'str' object has no attribute 'read'
```

### Operating System

Windows

### FastAPI Version

0.95.0

### Python Version

Python 3.10.10

**Comment by abhint** [comment]
https://fastapi.tiangolo.com/tutorial/request-files/?h=file+up#request-files

**Reply by plsholdmybeer** [reply]
did I miss something in the docs?
edit: I guess he's just farming replies :D

**Comment by Viicos** [comment]
https://github.com/tiangolo/fastapi/pull/9383/

**Reply by Kludex** [reply]
@Viicos Is this still an issue?

**Reply by Viicos** [reply]
Haven't tested in a while, the draft PR might still be relevant but I haven't got time to get back at it

**Comment by viertel97** [comment]
The error persists in the current (0.115.11) version when i use
`async def upload_file(classification_file: Annotated[UploadFile, File()], allowed_companies: Annotated[list, Form()]):` and then upload a file and send some strings. Its working if i switch the `Annotated[list, Form()]` to `Annotated[str, Form()]`

**Comment by VimalN2005** [comment]
Hi everyone,

I investigated this issue and found the root cause in `fastapi/dependencies/utils.py`.

### Root Cause
In `_extract_form_body`, for single bytes files, FastAPI properly checks `isinstance(value, UploadFile)` before calling `await value.read()`.
However, for sequence bytes files (`is_bytes_sequence_annotation`), the loop was unconditionally calling `await sub_value.read()`:
```python
for sub_value in value:
    results.append(await sub_value.read())

**Comment by Jeanlaris** [comment]
"Great breakdown of the root cause! To properly fix this in fastapi/dependencies/utils.py, we need a conditional safeguard inside the is_bytes_sequence_annotation block to ensure that sub_value is validated as a valid file/binary sequence before .read() is awaited. When Swagger UI sends an empty value/string, the type coercion needs an explicit type guard or fallback check to avoid calling .read() on primitive string instances:

Python
if sub_value and hasattr(sub_value, "read"):
    results.append(await sub_value.read())
This safeguards sequential byte unwrapping against empty form data inputs sent from the UI."

**Comment by VimalN2005** [comment]
Thanks @Jeanlaris! That's a great point.

Checking for `.read()` via duck typing alongside `isinstance(sub_value, UploadFile)` (which aligns with the single-file check on line 924) ensures we never attempt to call `.read()` on primitive string instances passed from Swagger UI or other clients:

```python
for sub_value in value:
    if isinstance(sub_value, UploadFile) or hasattr(sub_value, "read"):
        results.append(await sub_value.read())
    else:
        results.append(sub_value)

**Comment by VimalN2005** [comment]
Hi @YuriiMotov @tiangolo,

This bug has been analyzed and validated with @Jeanlaris in this thread. The fix and regression tests are fully verified on branch [`fix-list-bytes-empty-form-value`](https://github.com/VimalN2005/fastapi/tree/fix-list-bytes-empty-form-value).

Could you please give the green light so I can open the pull request?

**Reply by YuriiMotov** [reply]
Sorry, but we are not ready to work on this right now.
We will track this internally and take a look when we have time for it.
Thanks!

**Reply by Jeanlaris** [reply]
Understood, completely respect the maintainers' timeline and backlog priorities. 

I just want to express my genuine gratitude to Tiagolo and the entire FastAPI ecosystem. Coming from a simple background as a mechanical electrician working with coffee dryers in rural Brazil, without the traditional formal degrees that many of you have, having the opportunity to debate core architecture and learn at this level has been an incredible privilege. 

I'm just a guy typing code, learning fast, and deeply honored to contribute. Take all the time you need, and keep building amazing things!

**Comment by nanisadw3** [comment]
Hi! 👋

This occurs because Swagger UI sends an empty string `""` or `null` when "Send empty value" is selected for multi-part/form uploads or binary lists, which FastAPI tries to treat as a collection of binary chunks.

### How to handle this robustly:

Instead of bare `list[bytes]`, define the parameter as optional with an empty list default and use `UploadFile`:

```python
from typing import Annotated, List
from fastapi import FastAPI, File, UploadFile

app = FastAPI()

@app.post("/upload")
async def upload_files(
    files: Annotated[List[UploadFile], File()] = []
):
    if not files or (len(files) == 1 and files[0].filename == ""):
        return {"uploaded": 0, "message": "No files provided"}
    
    contents = [await f.read() for f in files]
    return {"uploaded": len(contents)}
```

Using `List[UploadFile]` prevents the internal `AttributeError` when Swagger UI passes an empty payload and gives you proper access to stream reading, file size validation, and empty-check handling.
